feat(http): Introduction of the `fetch` Backend for the `HttpClient`

JeanMeche opened the pull request:
This commit introduces a new `HttpBackend` implentation which makes requests using the fetch API

This feature is a developer preview and is opt-in.
It is enabled by setting the providers with `provideHttpClient(withFetch())`.

Currently the report progress is not implemented and using `reportProgress: true` will throw an error.  

NB: The fetch API is experimental on Node but available without flags from Node 18 onwards.
NB2: The tests provided in the PR won't run, as `fetch` is not available on the node version provided by the bazel workspace (`16.14`)

---- 
Edit: pinning a answer I gave bellow on the purpose of this new backend: 

`Xhr` is not supported natively on NodeJS an requires a polyfill. Currently Angular uses [xhr2](https://www.npmjs.com/package/xhr2) for this. However, the polyfill is side-effectful and doesn't work in non-Node.js environments (such as Edge Workers). There are also others issues like https://github.com/angular/angular/issues/46930 (its doesn't support Gzip).


## PR Type
What kind of change does this PR introduce?

- [x] Feature

## Does this PR introduce a breaking change?

- [x] No


Closes https://github.com/angular/angular/issues/27689

JeanMeche committed:
feat(http): Introduction of the `fetch` Backend for the `HttpClient`

This commit introduces a new `HttpBackend` implentation which makes requests using the fetch API

This feature is a developer preview and is opt-in.
It is enabled by setting the providers with `provideHttpClient(withFetch())`.

NB: The fetch API is experimental on Node but available without flags from Node 18 onwards.

alan-agius4 commented on packages/common/http/src/fetch.ts:
```diff
@@ -0,0 +1,197 @@
+
+  // stored as lowercase in the map
+  const xRequestUrl = 'X-Request-URL'.toLocaleLowerCase();
+  if (response.headers.has(xRequestUrl)) {
+    return response.headers.get(xRequestUrl) ?? undefined;
+  }
+  return undefined;
```
This can be simplified if you change the return type of the function.

```suggestion
return response.headers.get(xRequestUrl) 
```

ajafff commented on packages/common/http/src/fetch.ts:
```diff
@@ -0,0 +1,198 @@
+/**
+ * Abstract class to provide a mocked implementation of `fetch()`
+ */
+export abstract class FetchFactory {
```
should the class used as DI token really contain mock code? this code will always end up in production bundles, but is totally unused

ajafff commented on packages/common/http/src/provider.ts:
```diff
@@ -233,3 +235,15 @@ export function withRequestsMadeViaParent(): HttpFeature<HttpFeatureKind.Request
+ * @developerPreview
+ */
+export function withNativeFetch(): HttpFeature<any> {
+  return makeHttpFeature(6 as HttpFeatureKind, [
```
should this become a regular enum member?

ajafff commented on packages/common/http/src/fetch.ts:
```diff
@@ -0,0 +1,198 @@
+        const url = getResponseUrl(response) ?? req.url;
+
+        function sendResult(errorOrBody: any) {
+          // Normalize another potential bug (this one comes from CORS).
```
can you add a bit more context? it's not clear which bug you are refering to and how its different from a regular status 0 error

JeanMeche replied:
I remove that part, `fetch` does error on timeout.

alan-agius4 commented on packages/common/http/src/fetch.ts:
```diff
@@ -0,0 +1,198 @@
+        }
+
+        // The response body is an async stream, the Reponse API provides helpers
+        if (req.responseType === 'json') {
```
Switch seems like a better choice here.

alan-agius4 commented on packages/common/http/src/provider.ts:
```diff
@@ -233,3 +235,15 @@ export function withRequestsMadeViaParent(): HttpFeature<HttpFeatureKind.Request
+ * Configures the current `HttpClient` instance to make requests using the fetch API.
+ * @developerPreview
+ */
+export function withNativeFetch(): HttpFeature<any> {
```
Maybe remove the word 'native'?

alan-agius4 commented on packages/common/http/src/fetch.ts:
```diff
@@ -0,0 +1,198 @@
+        function sendResult(errorOrBody: any) {
+          // Normalize another potential bug (this one comes from CORS).
+          if (status === 0) {
+            status = !!errorOrBody ? HttpStatusCode.Ok : 0;
```
```suggestion
            status = errorOrBody ? HttpStatusCode.Ok : 0;
```

alan-agius4 commented on packages/common/http/src/fetch.ts:
```diff
@@ -0,0 +1,198 @@
+
+  handle(req: HttpRequest<any>): Observable<HttpEvent<any>> {
+    // Everything happens on Observable subscription.
+    return new Observable((observer: Observer<HttpEvent<any>>) => {
```
IMO, using an observer here does not seem needed and it also adds extra complexity and impedes readability. 

I'd prefer something like;
```ts
from(fetch(...))
  .pipe(
    mergeMap(async (response) => {
      const text = await response.text()
      ...
    }),
    startWith({type: HttpEventType.Sent}),
    finalize(() => /*  teardown / abort logic */ )
  );

```

JoostK replied:
Interesting, I prefer the current approach especially from a debugging/stacktrace point of view.

JeanMeche replied:
I'm open to both. Just to compare I pushed an alternative, more reactive, implementation. I can still revert if needed.

alan-agius4 replied:
What I like with my suggested approach is that it enables usability of async/await which is not possible with previous approach.

JeanMeche replied:
I'm sold, async/await makes it so much more readable !

JeanMeche replied:
I've changed it a bit again, to support the report progress. Let me know what you think.

alan-agius4 replied:
@JeanMeche fixups need to match a previous commit example.

```
fixup! feat(http): Introduction of the `fetch` Backend for the `HttpClient`
```

alan-agius4 reviewed (commented):
Couple of initial comments (This is not a full review).

JeanMeche commented on packages/common/http/src/fetch.ts:
```diff
@@ -0,0 +1,197 @@
+  public headers?: HeadersInit;
+  public credentials?: RequestCredentials;
+
+  private promise = new Promise<Response>((resolve, error) => {
```
That's just the mock implementation. It isn't used outside the tests

JeanMeche commented on packages/common/http/src/fetch.ts:
```diff
@@ -0,0 +1,197 @@
+/**
+ * Abstract class to provide a mocked implementation of `fetch()`
+ */
+export abstract class FetchFactory {
```
Yeah, it much cleaner now.

JeanMeche commented:
Also, @alan-agius4 pointed out one DX issue. On projects with multi `provideHttpClient`, it's easy to forget a  `withFetch()` call. 

I'll look for a way to provide a global use of the `FetchBackend`

alan-agius4 commented on packages/common/http/test/fetch_spec.ts:
```diff
@@ -0,0 +1,326 @@
+        return this.promise;
+      }
+
+  async mockFlush(status: number, statusText: string, body?: any): Promise<void> {
```
looks like `async` is not needed here.

alan-agius4 commented on packages/common/http/test/fetch_spec.ts:
```diff
@@ -0,0 +1,326 @@
+    this.resolve(response);
+  }
+
+  async mockErrorEvent(error: any) {
```
same.

alan-agius4 commented on packages/common/http/test/fetch_spec.ts:
```diff
@@ -0,0 +1,326 @@
+    this.error(error);
+  }
+
+  async mockAbortEvent() {
```
same

alan-agius4 commented on packages/common/http/test/fetch_spec.ts:
```diff
@@ -0,0 +1,326 @@
+export class MockFetchFactory extends FetchFactory {
+  public readonly response = new MockFetchResponse();
+  public readonly request = new MockFetchRequest();
+  protected resolve!: Function;
```
Looks like this and `error` should be private.

alan-agius4 commented on packages/common/http/src/fetch.ts:
```diff
@@ -0,0 +1,176 @@
+    req.headers.forEach((name, values) => (headers[name] = values.join(',')));
+
+    // Add an Accept header if one isn't present already.
+    if (!headers['Accept']) {
+      headers['Accept'] = 'application/json, text/plain, */*';
+    }
```
NIT picking 


```suggestion
    headers['Accept'] ??= 'application/json, text/plain, */*';
```

alan-agius4 commented on packages/common/http/src/fetch.ts:
```diff
@@ -0,0 +1,176 @@
+    }
+
+    // Auto-detect the Content-Type header if one isn't present already.
+    if (!headers['Content-Type']) {
+      const detectedType = req.detectContentTypeHeader();
+      // Sometimes Content-Type detection fails.
+      if (detectedType !== null) {
+        headers['Content-Type'] = detectedType;
+      }
+    }
```
NIT picking 
```suggestion
    headers['Content-Type'] ??= req.detectContentTypeHeader() ?? undefined;
```

JeanMeche replied:
headers does not accept `undefined` values.

alan-agius4 reviewed (dismissed):
Thanks @JeanMeche for this.

alan-agius4 commented on packages/common/http/src/fetch.ts:
```diff
@@ -0,0 +1,176 @@
+    const init = this.createRequestInit(req);
+    return from(this.fetchImpl(req.url, {signal: abortController.signal, ...init}))
+        .pipe(
+            catchError((error) => {
```
In case the request is aborted, we should throw an error, thus we need to filter here.

`!(error instanceof DOMException && error.name === 'AbortError')`

JeanMeche replied:
I'm not sure I understand what you mean.  
To me the implementations matches what we have with the `HttpXhrBackend` and is covered by this test : 

```typescript
    it('emits an error when browser cancels a request', done => {
      backend.handle(TEST_POST).subscribe({
        error: (err: HttpErrorResponse) => {
          expect(err instanceof HttpErrorResponse).toBe(true);
          expect(err.error instanceof DOMException).toBeTruthy();
          done();
        }
      });
      fetchMock.mockAbortEvent();
    });
```

Or did I miss something  ?

alan-agius4 replied:
Oh.. interesting. I missed that.

alan-agius4 commented on packages/common/http/src/fetch.ts:
```diff
@@ -0,0 +1,176 @@
+  handle(req: HttpRequest<any>): Observable<HttpEvent<any>> {
+    const abortController = new AbortController();
+    const init = this.createRequestInit(req);
+    return from(this.fetchImpl(req.url, {signal: abortController.signal, ...init}))
```
Actually, I think we might simply this even more... Gimme me 10 mins.

alan-agius4 replied:
What do you think?

```ts
  private async doRequest(req: HttpRequest<any>, signal: AbortSignal): Promise<HttpResponse<any>> {
    const init = this.createRequestInit(req);
    let response: Response | undefined;

    try {
      response = await this.fetchImpl(req.url, {signal, ...init});
    } catch (error: any) {
      // Filter for abort errors.
      throw new HttpErrorResponse({
        error,
        status: error.status ?? 0,
        statusText: error.statusText,
        url: req.url,
        headers: error.headers,
      });
    }

    let body;
    try {
      body = await this.getBody(req, response);
    } catch {
      // Body parsing error.
      throw new HttpErrorResponse({
        error,
        headers: new HttpHeaders(response.headers),
        status: response.status,
        statusText: response.statusText,
        url: getResponseUrl(response) ?? req.url,
      });
    }

    const headers = new HttpHeaders(response.headers);
    let status = response.status;
    // Normalize another potential bug (this one comes from CORS).
    if (status === 0) {
      status = body ? HttpStatusCode.Ok : 0;
    }

    const statusText = response.statusText;
    // Read the response URL from the response instance and fall back on the request URL.
    const url = getResponseUrl(response) ?? req.url;

    // ok determines whether the response will be transmitted on the event or
    // error channel. Unsuccessful status codes (not 2xx) will always be errors,
    // but a successful status code can still result in an error if the user
    // asked for JSON data and the body cannot be parsed as such.
    const ok = status >= 200 && status < 300;

    if (ok) {
      return new HttpResponse({
        body: body,
        headers,
        status,
        statusText,
        url,
      });
    }

    throw new HttpErrorResponse({
      error: body,
      headers,
      status,
      statusText,
      url,
    });
  }

  handle(req: HttpRequest<any>): Observable<HttpEvent<any>> {
    const abortController = new AbortController();
    return from(this.doRequest(req, abortController.signal)).pipe(
      startWith({type: HttpEventType.Sent} as HttpSentEvent),
      finalize(() => {
        // Aborting the fetch call when the observable is unsubscribed
        abortController.abort();
      })
    );
  }
```

JeanMeche replied:
The best of both worlds 🚀

HyperLife1119 commented:
The `HttpRequest#reportProgress` option is not available for the Fetch API, maybe we need to check for it and throw an exception?

alan-agius4 commented:
@HyperLife1119, we can explore/add progress in a follow up.

alan-agius4 commented:
By the way the commit scope should be `http` instead of `common`.

JiaLiPassion commented:
`zone.js` monkey patch `XMLHttpRequest` by default, and `xhr.send()` schedule a `macroTask` and the task finishs when `onload` event is triggered, with `fetch`, the zone.js `fetch patch` is not loaded by default, so only a `microTask` will be scheduled when `fetch().then()` is called, most use cases will not be impacted, but if some application monitor the `macroTask` with there own `ZoneSpec`, `FetchBackend` may introduce a break change.

AndrewKushnir commented:
@JiaLiPassion thanks for the comment 👍 What'd be the best way to handle that?

I'm wondering if something like this would work:
- We can update the docs of the `withFetch` function to mention that an extra Zone.js patch should be imported (@JiaLiPassion what'd be the right patch to load?).
- Within the `withFetch` function in **dev mode** we can check if Zone.js is enabled (check that it's not a NoopZone) in the app and if so, whether the patch is installed. If the patch is not there - we can log a warning into a console.

WDYT?

gilsdav commented:
Will Angular Universal will automatically support this ? Dors it réduire minimum node 18 ?

alan-agius4 commented:
> Will Angular Universal will automatically support this ? Dors it réduire minimum node 18 ?

Yes to both.

jnizet commented:
Hi.

I'd be interested to know what the rationale is behind this new HTTP backend. I've seen a disadvantage mentioned (no support for progress), but I haven't seen any advantage/benefit mentioned. 

- Is it more efficient somehow?
- Does it reduce the bundle size significantly?
- Is it easier (or harder) to mock with e2e testing frameworks?
- Does it provide additional features compared to the XHR backend?

JeanMeche commented:
> Hi.
> 
> I'd be interested to know what the rationale is behind this new HTTP backend. I've seen a disadvantage mentioned (no support for progress), but I haven't seen any advantage/benefit mentioned.
> 
> * Is it more efficient somehow?
> * Does it reduce the bundle size significantly?
> * Is it easier (or harder) to mock with e2e testing frameworks?
> * Does it provide additional features compared to the XHR backend?

Xhr are not supported natively on NodeJS an requires a polyfill. Currently Angular uses [https://www.npmjs.com/package/xhr2](`xhr2`) for this.  However,  the polyfill is side-effectful and doesn't work in non-Node.js environments (such as Edge Workers). There are also others issues like #46930 (its doesn't support Gzip).

JeanMeche commented:
> `zone.js` monkey patch `XMLHttpRequest` by default, and `xhr.send()` schedule a `macroTask` and the task finishs when `onload` event is triggered, with `fetch`, the zone.js `fetch patch` is not loaded by default, so only a `microTask` will be scheduled when `fetch().then()` is called, most use cases will not be impacted, but if some application monitor the `macroTask` with there own `ZoneSpec`, `FetchBackend` may introduce a break change.

@JiaLiPassion How can we detect if `zone-patch-fetch` has been loaded. ?

alan-agius4 commented on packages/common/http/src/fetch.ts:
```diff
@@ -0,0 +1,240 @@
+      }
+
+      // retreive first chunk
+      reader.read().then(handleChunk).then((finalContent: Uint8Array) => {
```
This is a floating promise. We should avoid these as these cause problem on Node.

JeanMeche commented on packages/common/http/src/fetch.ts:
```diff
@@ -0,0 +1,240 @@
+
+  handle(req: HttpRequest<any>): Observable<HttpEvent<any>> {
+    const abortController = new AbortController();
+    return from(this.doFetch(req, abortController.signal))
```
@alan-agius4 Your last edit went a bit too far I think. The subject needs to complete for the merge to complete.

alan-agius4 replied:
Oh right. That should trivial to solve. 

I am currently AFK, but can look at it on Monday.

alan-agius4 commented on packages/common/http/src/fetch.ts:
```diff
@@ -0,0 +1,229 @@
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import {inject, Injectable} from '@angular/core';
+import {BehaviorSubject, from, merge, Observable} from 'rxjs';
+import {finalize, map, mergeMap, startWith, tap} from 'rxjs/operators';
+
+import {HttpBackend} from './backend';
+import {HttpHeaders} from './headers';
+import {HttpRequest} from './request';
+import {HttpDownloadProgressEvent, HttpErrorResponse, HttpEvent, HttpEventType, HttpHeaderResponse, HttpResponse, HttpSentEvent, HttpStatusCode,} from './response';
```
```suggestion
import {inject, Injectable} from '@angular/core';
import {BehaviorSubject, from, merge, Observable} from 'rxjs';
import {finalize, map} from 'rxjs/operators';

import {HttpBackend} from './backend';
import {HttpHeaders} from './headers';
import {HttpRequest} from './request';
import {HttpDownloadProgressEvent, HttpErrorResponse, HttpEvent, HttpEventType, HttpHeaderResponse, HttpResponse, HttpStatusCode} from './response';
```

alan-agius4 commented on packages/common/http/src/fetch.ts:
```diff
@@ -0,0 +1,229 @@
+      const chunks: Uint8Array[] = [];
+      const reader = response.body.getReader();
+      let receivedLength = 0;
+
+      if (request.reportProgress) {
+        eventStream.next(new HttpHeaderResponse({headers, status, statusText, url}));
+      }
```
This should probably be moved out of the `response.body` block.

```suggestion

      if (request.reportProgress) {
        eventStream.next(new HttpHeaderResponse({headers, status, statusText, url}));
      }
```

alan-agius4 commented on packages/common/http/src/fetch.ts:
```diff
@@ -0,0 +1,229 @@
+    };
+  }
+
+  private concatChunks(chunks: Uint8Array[], totalLength: number) {
```
Can you please add a explicate return type?

alan-agius4 commented on packages/common/http/src/fetch.ts:
```diff
@@ -0,0 +1,229 @@
+    try {
+      switch (request.responseType) {
+        case 'json':
+          // Calling `text()` instead of `json()` to allow XSSI stripping
```
Outdated comment.

devversion commented on WORKSPACE:
```diff
@@ -64,7 +64,7 @@ load("@rules_nodejs//nodejs:repositories.bzl", "nodejs_register_toolchains")
 nodejs_register_toolchains(
     name = "nodejs",
-    node_version = "16.14.0",
+    node_version = "16.15.0",
```
If we bump this, we need to also update the CircleCI version

JeanMeche replied:
I was just about to push this change 😊 👍

Also Paul, maybe you could help me find where I could add the `--experimental-fetch` which is needed by node to fix the tests. 

Thx in advance !

devversion commented on packages/common/http/test/fetch_spec.ts:
```diff
@@ -0,0 +1,427 @@
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import {HttpEvent, HttpEventType, HttpRequest, HttpResponse} from '@angular/common/http';
```
If you need the experimental fetch flag for this spec, then the owning `jasmine_node_test` can have a `templated_args = ["--node-options=--experimental-fetch=true"]`. Not sure if it's exactly that API, but something around these lines should work.

I think we'd want a separate `jasmine_node_test` target for this spec, so that we don't enable the flag for all other tests.

alan-agius4 replied:
You almost got the API right @devversion 


```

templated_args = ["--node-options=--experimental-fetch"]
```

alan-agius4 replied:
**Update:** seeing that we had to disable some tests, I’d prefer if we use a Node 18 tool chain and disable RBE for this target.

Potentially, have it as a different build target if @devversion prefers it that way.

(I can do this)

JeanMeche replied:
FYI, I tried adding `templated_args = ["--node-options=--experimental-fetch"]`  to `jasmine_node_test ` in `packages/common/http/test/BUILD.bazel` but the doesn't enable the experimental fetch flag.

devversion replied:
> What’s the reason why you’d like a different target? There is no downside to enable it for the entire target

@alan-agius4 the downside would be that other parts of the framework _could_ potentially rely on these experimental APIs and we wouldn't notice. A separate target minmizes this and makes it explicit.

> Update: seeing that we had to disable some tests, I’d prefer if we use a Node 18 tool chain and disable RBE for this target.

I'm good with that as well.

JeanMeche replied:
@alan-agius4 It works 👌👌. 

It even raised a broken test in another package which should be fixed now !

AndrewKushnir commented on packages/common/http/src/fetch.ts:
```diff
@@ -0,0 +1,230 @@
+  if (response.url) {
+    return response.url;
+  }
+  // stored as lowercase in the map
+  const xRequestUrl = 'X-Request-URL'.toLocaleLowerCase();
```
nit: we can probably extract this outside of the function as well (similar to `XSSI_PREFIX` const).

AndrewKushnir commented on packages/common/http/src/fetch.ts:
```diff
@@ -0,0 +1,230 @@
+/**
+ * Uses `fetch` to send requests to a backend server.
+ *
+ * This `FetchBackend` requires the support of the Fetch API which is available on all evergreen
```
nit: we can add a link here:
```suggestion
 * This `FetchBackend` requires the support of the [Fetch API](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API) which is available on all evergreen
```

AndrewKushnir commented on packages/common/http/src/fetch.ts:
```diff
@@ -0,0 +1,230 @@
+ * browsers and on NodeJS from v18 onward.
+ *
+ * @see {@link HttpHandler}
+ *
+ * @developerPreview
```
```suggestion
 *
 * @publicApi
 * @developerPreview
```

AndrewKushnir commented on packages/common/http/src/fetch.ts:
```diff
@@ -0,0 +1,230 @@
+    const url = getResponseUrl(response) ?? request.url;
+
+    let status = response.status;
+    let body: any = null;
```
Let's may be leave a quick comment in the code on why we use `any`?

alan-agius4 replied:
I’d prefer to have explicate types here.

JeanMeche replied:
The issue is that it will be narrowed to `any` because of `JSON.parse`. 

Do you really prefere `string | ArrayBuffer | Blob | null | any` over `any` ?

alan-agius4 replied:
`string | ArrayBuffer | Blob | object | null` maybe?

I also believe `JSON.parse` would need to be casted as in G3 it is not allowed to have untyped `JSON.parse`

AndrewKushnir commented on packages/common/http/src/request.ts:
```diff
@@ -113,6 +113,8 @@ export class HttpRequest<T> {
    *
    * Progress events are expensive (change detection runs on each event) and so
    * they should only be requested if the consumer intends to monitor them.
+   *
+   * This option is not supported when using the `FetchBackend`.
```
It looks like we've added support for it now, so we can remove this comment?

AndrewKushnir commented on packages/common/http/src/provider.ts:
```diff
@@ -233,3 +236,24 @@ export function withRequestsMadeViaParent(): HttpFeature<HttpFeatureKind.Request
+ *
+ * This `FetchBackend` requires the support of the Fetch API which is available on all evergreen
+ * browsers and on NodeJS from v18 onward.
+ *
+ * @developerPreview
```
```suggestion
 *
 * @publicApi
 * @developerPreview
```

AndrewKushnir reviewed (commented):
@JeanMeche huge thanks for the PR, it looks great 👍 I've left some minor comments only.

AndrewKushnir reviewed (commented):
@JeanMeche huge thanks for the PR, it looks great 👍 I've left some minor comments only.

alan-agius4 commented on packages/common/http/src/fetch.ts:
```diff
@@ -0,0 +1,230 @@
+    });
+  }
+
+  private parseBody(request: HttpRequest<any>, response: Response, binContent: Uint8Array): any {
```
Can we add an explicate return type please?

JeanMeche commented:
> The `HttpRequest#reportProgress` option is not available for the Fetch API, maybe we need to check for it and throw an exception?

@HyperLife1119  Download report progress has been added. Upload progress isn't possible with fetch though.

ajafff commented on packages/common/http/src/provider.ts:
```diff
@@ -233,3 +236,27 @@ export function withRequestsMadeViaParent(): HttpFeature<HttpFeatureKind.Request
+ * @developerPreview
+ */
+export function withFetch(): HttpFeature<any> {
+  if ((typeof ngDevMode === 'undefined' || ngDevMode) && typeof fetch !== 'function') {
```
this is not correct if someone provides a FetchFactory with, e.g. `undici`

JeanMeche replied:
The `FetchFactory` is not public.

ajafff commented on packages/common/http/src/fetch.ts:
```diff
@@ -0,0 +1,238 @@
+            total: contentLength ? +contentLength : undefined,
+            loaded: receivedLength,
+            partialText: request.responseType === 'text' ?
+                new TextDecoder().decode(this.concatChunks(chunks, receivedLength)) :
```
this could become very expensive to concat and decode everything received so far for each chunk.
consider reusing the TextDecoder and calling decode with the `stream` option enabled

ajafff commented on packages/common/http/src/fetch.ts:
```diff
@@ -0,0 +1,238 @@
+            type: HttpEventType.DownloadProgress,
+            total: contentLength ? +contentLength : undefined,
+            loaded: receivedLength,
+            partialText: request.responseType === 'text' ?
```
I really hope we can have something similar for `responseType === 'arraybuffer'` in the future

JeanMeche replied:
You could open a feature request for this.

ajafff commented on packages/common/http/src/fetch.ts:
```diff
@@ -0,0 +1,238 @@
+    const eventStream = new BehaviorSubject<HttpEvent<any>>({type: HttpEventType.Sent});
+
+    return merge(
+               from(this.doRequest(req, abortController.signal, eventStream.next).finally(() => {
```
passing an unbound `BehaviorSubject#next` might work in rxjs@6, but throws in rxjs@7

JeanMeche replied:
I don't have issues to report on my tests here : https://stackblitz.com/edit/angular-fetch-backend

Can you elaborate?

JeanMeche replied:
@alan-agius4 Do we have a way to run tests with RxJS 7 ?

alan-agius4 replied:
Not the unit test, but G3 is on RXJS 7

alan-agius4 replied:
Would suggest the below and maybe add a comment to why it is needed.

```
this.doRequest(req, abortController.signal, (e) => eventStream.next(e))
```

Or 

```
this.doRequest(req, abortController.signal, eventStream.next.bind(eventStream))
```

JeanMeche replied:
Already done 😉

ajafff commented on packages/common/http/src/fetch.ts:
```diff
@@ -0,0 +1,238 @@
+      let receivedLength = 0;
+
+      while (true) {
+        const {done, value} = await reader.read();
```
can we avoid using the reader when nobody cares about download progress? I guess the browser is better at creating the final buffer

alan-agius4 replied:
We can benchmark, but I doubt there would be any noticeable performance difference.

alxhub replied:
One downside is that if the response takes a few packets to deliver, then each promise returned by `reader.read()` will trigger a new round of change detection (via zone.js) even though we're not doing anything besides accumulating into the chunk buffer.

alan-agius4 commented on packages/common/http/src/headers.ts:
```diff
@@ -258,6 +249,24 @@ export class HttpHeaders {
     }
   }
 
+  private allocateHeaderEntries(name: string, values: any) {
```
```suggestion
  private allocateHeaderEntries(name: string, values: unknown): void {
```

alan-agius4 replied:
Also, maybe consider `set` instead of `allocate`? It's more explicate and shorter.

alan-agius4 commented on packages/common/http/src/headers.ts:
```diff
@@ -258,6 +249,24 @@ export class HttpHeaders {
   }
 
+  private allocateHeaderEntries(name: string, values: any) {
+    let headerValues: string[];
+
+    if (typeof values === 'string') {
+      headerValues = [values];
+    } else if (typeof values === 'number') {
+      headerValues = [values.toString()];
+    } else {
+      headerValues = values.map((value: any) => value.toString());
+    }
+
+    if (headerValues.length > 0) {
+      const key = name.toLowerCase();
+      this.headers.set(key, headerValues);
+      this.maybeSetNormalizedName(name, key);
+    }
+  }
```
This can be simplified slightly, because;
1. The array will always have an item or more. 
2. No need to check for number/string.

```suggestion
const headerValues = (Array.isArray(values) ? values : [values]).map((value) => value.toString();
const key = name.toLowerCase();
this.headers.set(key, headerValues);
this.maybeSetNormalizedName(name, key);
```

ajafff commented on packages/common/http/src/fetch.ts:
```diff
@@ -0,0 +1,243 @@
+            type: HttpEventType.DownloadProgress,
+            total: contentLength ? +contentLength : undefined,
+            loaded: receivedLength,
+            partialText: request.responseType === 'text' ? decoder!.decode(value, {stream: true}) :
```
the latest change here changed the behavior. `TextDecoder#decode` only returns the decoded string for the current chunk. before the change this always contained the whole text received so far. that means the previous decode result needs to be kept and concatenated with the new one

ajafff commented on packages/common/http/src/fetch.ts:
```diff
@@ -0,0 +1,243 @@
+  private readonly fetchImpl =
+      inject(FetchFactory, {optional: true})?.fetch ?? fetch.bind(globalThis);
+
+  handle(req: HttpRequest<any>): Observable<HttpEvent<any>> {
```
is this method guaranteed to be called for each subscription? if not, this whole method body should be wrapped in `defer(() => { /* ... */ })`. a previous implementation did something similar by returning `new Observable()`.

I just want to make sure that multiple subscriptions to the same request Observable don't affect each other:

```ts
const r = http.get('.');
r.subscribe().unsubscribe();
r.subscribe(console.log); // should still work and not reuse the abortController of the first request
```

JeanMeche replied:
I'm not sure I understand your wories here but here is the call site: 

https://github.com/angular/angular/blob/85b4941be137a2fcdc664dc870e408dd72ad7de7/packages/common/http/src/client.ts#L542-L543

ajafff replied:
XhrBackend contains code exactly for this purpose: https://github.com/angular/angular/blob/85b4941be137a2fcdc664dc870e408dd72ad7de7/packages/common/http/src/xhr.ts#L69-L70

based on your comment it seems that is no longer necessary or at least the comment is outdated

alxhub replied:
This is still necessary.

`HttpClient` may lazily call `handle`, but in the middle of the chain any interceptor is also free to create many intermediate request `Observable`s, and may not subscribe to them. Just calling `handle` shouldn't trigger a request.

JeanMeche replied:
Thx for the insight Alex, @alan-agius4 that leaves us with no other choice than a floating promise right ? (cf my latest commit)

alan-agius4 replied:
Well, you can refactor the code to avoid using the observer.

I think something like the below should be enough
```ts
merge(
  of(null).pipe(
    switchMap(() => this.doRequest(req, abortController.signal, (event) => eventStream.next(event)).finally(() => eventStream.complete()))
  ),
  eventStream,
)
```

alxhub replied:
Floating promises are normal and reasonable. Under the hood, `switchMap` is going to take the promise returned by `doRequest` and do `.then(...)` to emit into the observer when it completes (basically, it creates a floating promise).

alan-agius4 committed:
build: run `//packages/common/http/test` on Node.js 18

This commits adds configures `//packages/common/http/test` to use Node.js 18 toolchain which is needed to test the fetch implementation which use Node.js 18 APIs.

We also disable RBE for this target as it doesn't work with Node.js 18 right now. See https://github.com/angular/dev-infra/issues/1017

alan-agius4 commented on packages/common/http/src/fetch.ts:
```diff
@@ -0,0 +1,248 @@
+ *
+ * This `FetchBackend` requires the support of the
+ * [Fetch API](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API) which is available on all
+ * evergreen browsers and on NodeJS from v18 onward.
```
```suggestion
 * evergreen browsers and on Node.js from v18 onward.
```

alan-agius4 commented on packages/common/http/src/headers.ts:
```diff
@@ -66,28 +66,20 @@ export class HttpHeaders {
           }
         });
       };
+    } else if (headers.constructor?.name === 'Headers') {
```
Use `headers instanceof Headers`. Although I am a bit confused as this method does not accept `Headers`

JeanMeche replied:
I intentionally removed `Headers` from the signature because this code exists outside the new `FetchBackend`.  
`instanceof Headers` just fails when run in a node 16 environment. 

The alternative would to move all the `Headers` related code to `FetchBackend`.

alan-agius4 commented on packages/common/http/src/headers.ts:
```diff
@@ -66,28 +66,20 @@ export class HttpHeaders {
         });
       };
+    } else if (typeof Headers !== 'undefined' && headers instanceof Headers) {
+      // TODO: Replace the condition with `instanceof Headers` when Node 16 support is dropped.
```
Comment is outdated.

alan-agius4 commented on packages/common/http/src/fetch.ts:
```diff
@@ -0,0 +1,256 @@
+    // Workaround: We need to schedule a macrotask to keep the NgZone unstable during the request
+    // This is due to the ZoneJS only patching patching the initial fetch promise (the header one)
+    // but not the body
+    const macroTaskCanceller = createBackgroundMacroTask();
```
With the changes in https://github.com/angular/angular/pull/50425/commits/1467946527d1c378a2bc7c605e18ee0e77ee49c2 we can remove `createBackgroundMacroTask`.

alan-agius4 commented on packages/common/http/src/fetch.ts:
```diff
@@ -0,0 +1,255 @@
+    // We can't use it now because from() doesn't accept generators on RxJS 6.
+    // Just replace `reportEvent` calls with `yield`.
+    const eventStream = new BehaviorSubject<HttpEvent<any>>({type: HttpEventType.Sent});
+    return merge(
```
For visibility as it's hard to find comments.


I think something like the below should be enough (I did not test it, just pseudo code)
```ts
merge(
  of(null).pipe(
    mergeMap(() => {
       return this.doRequest(req, abortController.signal, (event) => eventStream.next(event))
                .finally(() => eventStream.complete())
    })
  ),
  eventStream,
)
```

JeanMeche commented on packages/common/http/src/fetch.ts:
```diff
@@ -58,20 +57,14 @@ export class FetchBackend implements HttpBackend {
-                           .finally(() => {
-                             eventStream.complete();
-                           })}),
+               of(null).pipe(mergeMap(() => {
```
Oh, now I fully understand your earlier comment. Good catch and thank you for your help. I've added a corresponding unit test to cover that case !

alan-agius4 reviewed (approved):
LGTM

AndrewKushnir reviewed (commented):
@JeanMeche LGTM, just a couple comments.

Regarding the `createBackgroundMacroTask` - it looks like a leftover, so we can revert it? It'd be also helpful to rebase the PR on top of the most recent revision in the main branch.

AndrewKushnir reviewed (approved):
@JeanMeche looks great! Once again, huge thanks for the help with this feature 👍

AndrewKushnir reviewed (approved):
Reviewed-for: public-api

alxhub reviewed (approved):
Reviewed-for: public-api

The node version commit should be before the HTTP commit, so that the test doesn't fail on that commit.

AndrewKushnir commented:
> The node version commit should be before the HTTP commit, so that the test doesn't fail on that commit.

@JeanMeche could you please also merge all fixup commits into the main commit while reordering them based on the feedback above? Thank you.

AndrewKushnir committed:
fixup! feat(http): Introduction of the `fetch` Backend for the `HttpClient`

AndrewKushnir commented on packages/common/http/src/fetch.ts:
```diff
@@ -0,0 +1,251 @@
+      });
+    }
+
+    throw new HttpErrorResponse({
```
FYI, I've pushed a fixup commit: https://github.com/angular/angular/pull/50247/commits/ca57fd38d82aacf833ba6edc3b89e40ebd5bdcf1, please let me know what you think.

AndrewKushnir replied:
@JeanMeche FYI we've made some additional refactoring in https://github.com/angular/angular/pull/50247/commits/fcac36fb4ef4c8f912a9df17131f177d191279c8

**No actions are required at this moment**. We'll be running some tests in Google's codebase and proceed with the merge if those tests pass.

AndrewKushnir commented:
[Presubmit #2](http://test/OCL:538612988:BASE:538613030:1686177989391:5e29c329).

AndrewKushnir commented:
[Presubmit #3](http://test/OCL:538612988:BASE:538638228:1686184648337:8479d2eb).

AndrewKushnir commented:
[Presubmit](http://test/OCL:538612988:BASE:538658096:1686191620980:f580a92e).

AndrewKushnir commented:
[TGP](https://fusion2.corp.google.com/presubmit/538612988/OCL:538612988:BASE:538675181:1686197744188:b460fdaf/targets).

AndrewKushnir commented:
**Caretaker note**: TGP looks good, current set of approvals should be sufficient to proceed with the merge.

// cc @alxhub

alxhub commented:
This PR was merged into the repository by commit 85c54275825a57fd3c7055a99e58bb211e085af9.
